File: apps/api/src/myetal_api/schemas/works.py
"""Pydantic schemas for /me/works/* — the personal works library.

Returns the global Paper alongside the per-user UserPaper metadata
(added_via, added_at, hidden_at) so the client has everything in one
response without follow-up requests.
"""

from __future__ import annotations

import uuid
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field

from myetal_api.models import PaperSource, UserPaperAddedVia


class AddWorkRequest(BaseModel):
    """Body for POST /me/works.

    `identifier` accepts any DOI form Crossref understands — bare,
    `doi:`-prefixed, or full https://doi.org/... URL. Normalised inside
    `services/papers.py:normalise_doi`.
    """

    identifier: str = Field(min_length=1, max_length=500)


class PaperOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    doi: str | None
    openalex_id: str | None
    title: str
    subtitle: str | None
    authors: str | None
    year: int | None
    venue: str | None
    abstract: str | None
    url: str | None
    pdf_url: str | None
    image_url: str | None
    source: PaperSource


class WorkResponse(BaseModel):
    """A library entry: the global paper plus this user's per-entry fields.

    Returned by POST /me/works (single) and GET /me/works (list).
    """

    paper: PaperOut
    added_via: UserPaperAddedVia
    added_at: datetime
    hidden_at: datetime | None

File: apps/web/src/types/works.ts
/**
 * Shapes for /me/works endpoints — the personal works library.
 * Mirrors the backend WorkResponse / PaperOut / AddWorkRequest schemas.
 */

export type PaperSource = 'orcid' | 'crossref' | 'openalex' | 'manual';
export type UserPaperAddedVia = 'orcid' | 'manual' | 'share';

export interface PaperOut {
  id: string;
  doi: string | null;
  openalex_id: string | null;
  title: string;
  subtitle: string | null;
  authors: string | null;
  year: number | null;
  venue: string | null;
  abstract: string | null;
  url: string | null;
  pdf_url: string | null;
  image_url: string | null;
  source: PaperSource;
}

export interface WorkResponse {
  paper: PaperOut;
  added_via: UserPaperAddedVia;
  added_at: string;
  hidden_at: string | null;
}

export interface AddWorkRequest {
  identifier: string;
}

/** POST /me/works/sync-orcid response shape. */
export interface OrcidSyncResponse {
  added: number;
  updated: number;
  unchanged: number;
  skipped: number;
  errors: string[];
  /**
   * Populated only on the user's first successful ORCID sync, when the
   * server pre-built a draft "Publications" share. Null on every
   * subsequent sync. Used by the library page to emit the
   * `orcid_auto_draft_created` PostHog event + by the dashboard banner
   * to deep-link "Review draft".
   */
  auto_draft_share_id: string | null;
  auto_draft_paper_count: number | null;
}
